File: sdk-kotlin/src/main/kotlin/com/hobostays/rtdb/core/Tree.kt
package com.hobostays.rtdb.core

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject

/** The stored unit: a flattened leaf path and its value. */
data class Leaf(val path: String, val value: JsonElement)

/**
 * Flatten a value into leaves (§1), the port of WP1's `src/storage/tree.ts`.
 *  - `null` and `{}` produce NO leaves — both mean "nothing is stored here";
 *  - arrays are opaque leaf values, Firebase-style.
 *
 * Unlike the server's, this one does not validate keys or count leaves: §9's caps are the gateway's
 * to enforce (it answers TOOBIG), and the write API validates the leaf paths it is about to send.
 */
fun flatten(path: String, value: JsonElement): List<Leaf> {
    val leaves = mutableListOf<Leaf>()

    fun walk(at: String, node: JsonElement) {
        when {
            node is JsonObject -> for ((key, child) in node) walk(joinPath(at, key), child)
            node is JsonNull -> Unit // a null anywhere in the value means "no leaf here"
            else -> leaves += Leaf(at, node)
        }
    }

    walk(path, value)
    return leaves
}

/** Rebuild the value rooted at `root` from its leaves. No leaves -> null (§1). */
fun unflatten(root: String, leaves: List<Leaf>): JsonElement {
    if (leaves.isEmpty()) return JsonNull
    // A leaf sitting exactly at the root means the root itself is a scalar or an array.
    leaves.firstOrNull { it.path == root }?.let { return it.value }

    val tree = LinkedHashMap<String, Any>()
    for (leaf in leaves) {
        val segs = segments(relativePath(root, leaf.path))
        var node = tree
        for (i in 0 until segs.size - 1) {
            val existing = node[segs[i]]
            @Suppress("UNCHECKED_CAST")
            node = if (existing is LinkedHashMap<*, *>) {
                existing as LinkedHashMap<String, Any>
            } else {
                LinkedHashMap<String, Any>().also { node[segs[i]] = it }
            }
        }
        node[segs.last()] = leaf.value
    }
    return toJson(tree)
}

private fun toJson(node: Any): JsonElement = when (node) {
    is LinkedHashMap<*, *> -> JsonObject(node.entries.associate { (k, v) -> k as String to toJson(v!!) })
    else -> node as JsonElement
}

File: src/storage/tree.ts
import type { Json } from '../protocol/frames.ts';
import type { Limits } from '../protocol/limits.ts';
import { joinPath, relativePath, segments, validateSegment } from '../protocol/path.ts';

/** The stored unit: a flattened leaf path and its value (PROJECT_STATE: "flattened leaf paths"). */
export interface Leaf {
  path: string;
  value: Json;
}

export type FlattenResult = { ok: true; leaves: Leaf[] } | { ok: false; msg: string; tooBig?: true };

/**
 * Flatten a write's value into leaves (§1).
 *  - `null` and `{}` produce NO leaves — both mean "nothing is stored here" ("empty objects are
 *    never stored", and null deletes). The caller treats an empty result as a delete of the subtree.
 *  - arrays are opaque leaf values, Firebase-style.
 *  - object keys are validated as single path segments; a key with a `/` or a `.` is BADPATH, not
 *    silent nesting.
 * Stops as soon as MAX_LEAVES_PER_WRITE is exceeded, so an oversized write costs bounded work (§9).
 */
export function flatten(path: string, value: Json, limits: Limits): FlattenResult {
  const leaves: Leaf[] = [];

  const walk = (at: string, v: Json): string | null => {
    if (v !== null && typeof v === 'object' && !Array.isArray(v)) {
      for (const [key, child] of Object.entries(v)) {
        const keyErr = validateSegment(key);
        if (keyErr) return `${keyErr} at "${at}"`;
        const err = walk(joinPath(at, key), child as Json);
        if (err) return err;
      }
      return null;
    }
    if (v === null) return null; // a null anywhere in the value means "no leaf here"
    if (leaves.length >= limits.MAX_LEAVES_PER_WRITE) return 'TOOBIG';
    leaves.push({ path: at, value: v });
    return null;
  };

  const err = walk(path, value);
  if (err === 'TOOBIG') {
    return { ok: false, msg: `write exceeds ${limits.MAX_LEAVES_PER_WRITE} leaves`, tooBig: true };
  }
  if (err) return { ok: false, msg: err };
  return { ok: true, leaves };
}

/** Rebuild the JSON value rooted at `root` from its leaves. No leaves -> null (§1). */
export function unflatten(root: string, leaves: Leaf[]): Json {
  if (leaves.length === 0) return null;
  // A leaf sitting exactly at the root means the root itself is a scalar or an array.
  const exact = leaves.find((l) => l.path === root);
  if (exact) return exact.value;

  const out: { [k: string]: Json } = {};
  for (const leaf of leaves) {
    const segs = segments(relativePath(root, leaf.path));
    let node = out;
    for (let i = 0; i < segs.length - 1; i++) {
      const key = segs[i] as string;
      const next = node[key];
      if (next === undefined || next === null || typeof next !== 'object' || Array.isArray(next)) {
        node = node[key] = {};
      } else {
        node = next as { [k: string]: Json };
      }
    }
    node[segs[segs.length - 1] as string] = leaf.value;
  }
  return out;
}
